Sources/RedCat/Observers.swift

```diff
@@ -8,58 +8,34 @@
 import Foundation
 
 
-
 @usableFromInline
-class Observers {
-    
-    //faster than reading from dict
-    private weak var firstObserver : StoreDelegate?
-    private var otherObservers : [UUID : WeakDelegate] = [:]
+class Observers<State> {
     
-    @usableFromInline
-    func addObserver(_ observer: StoreDelegate) {
-        
-        if firstObserver == nil {
-            firstObserver = observer
-        }
-        else {
-            otherObservers[UUID()] = WeakDelegate(delegate: observer)
-        }
-    }
+    private var observers: [UUID: AnyStoreDelegate<State>] = [:]
     
     @usableFromInline
-    func notifyAll() {
-        
-        firstObserver?.storeWillChange()
-        
-        for observer in otherObservers.values {
-            observer.delegate?.storeWillChange()
-        }
-        
-        let invalidObservers = otherObservers.compactMap{key, value in
-            value.delegate == nil ? key : nil
-        }
-        
-        for key in invalidObservers {
-            otherObservers.removeValue(forKey: key)
-        }
-        
-        if
-            firstObserver == nil,
-            let (key, value) = otherObservers.first {
-            otherObservers.removeValue(forKey: key)
-            firstObserver = value.delegate
-        }
-        
+    func addObserver<S: StoreDelegate>(_ observer: S) -> StoreUnsubscriber where S.State == State {
+        let id = UUID()
+				observers[id] = AnyStoreDelegate(observer)
+				return StoreUnsubscriber { self.observers[id] = nil }
     }
     
+		@usableFromInline
+		func notifyAll(old: State, new: State, action: ActionProtocol) {
+				for observer in observers.values {
+					observer.storeDidChange(oldState: old, newState: new, action: action)
+				}
+		}
 }
 
-
```

Comment:
The intention to use a more passive model of unsubscribing was to avoid unsubscribe boilerplate in ```deinit```s or having to declare a ```Set<AnyCancellable>``` and store everything in there. UIKit did this a lot with its weak delegates. I'm willing to accept this source breaking change, if you can tell me how I can make this as simple as 
```
store.delegate = self
```
in UIKit without Combine.

Reply:
I would like to be able to control the life of the subscription, not only the long life of the subscriber (for example for a combine publishers)
It possible subscribe to the life of an object without Combine using obj_setAssociatedObject (I don't know it works for Linux) then there is no need to check the life of subscribers every time

Reply:
I see where you're coming from, I would just like to see a best-of-both-worlds approach that ideally isn't source breaking (even though this framework seems to have virtually no users yet, so nobody is really affected).

setAssociatedObject sounds like replacing one evil for another? Would the subscriber have to call this or the store/publisher?

Reply:
Observers could call this
but I think it might be better to return the original version of the `Observers` and add to them `StoreUnsubscriber` as `@discardableResult`

Reply:
That's what I was thinking. We wouldn't get rid of the lifetime checks then though, and we would continue to force subscribers to be objects.
Suggestion: there could be two overloads to subscribe to the store's publisher, one for lifetime-checked objects and one for combine subscribers; both get notified when ```notifyAll``` gets called.